rust-symcrypt/src/gcm.rs

```diff
@@ -4,97 +4,119 @@ use crate::block_ciphers::*;
 use crate::errors::SymCryptError;
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
 use std::vec;
 use symcrypt_sys;
 
 pub struct GcmState {
-    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    inner: Pin<Box<GcmStateInner>>,
+}
+
+struct GcmStateInner {
     state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
 }
 
-// Have to Box<> since memory address for Self is moved around when returning from GcmState::new()
-// My guess is that it has to do with Result<> wrapping the Self and SymCryptError.
-// The only accepted Cipher for GCM is AesBlock.
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime.
+/// Using an inner GCM state that is Pin<Box<T>> since the memory address for Self is moved around when returning from GcmState::new()
+/// My guess is that it has to do with Result<> wrapping the Self and SymCryptError.
```

Comment:
I'm not sure if it's due to the `Result` type, but rather just because `instance` is allocated on the stack, so when the function returns, the values have to be copied to their destination. In some cases the compiler might be smart enough to optimize this out, but it probably can't do so all the time.

(That said, standard caveat that I don't know much about Rust.)

Reply:
I had the same idea as well, when I tired this on other code like HMAC, there was no issue with the magic. The magic issue only started occurring once I introduced the Result type. Its entirely possible that we were just getting lucky before and the compiler was able to do some in line move to make the memory address match up.

I will remove my speculation for clarity